### PR TITLE
P4-1560 Add allocation filter to moves endpoint

### PR DESCRIPTION
**Pull request by cwrw**
To filter on moves with an allocation association, add a new filter
to either return all moves with an allocation if set to true, or to exclude
all moves associated to an allocation if set to false

### Jira link

P4- todo: create ticket 🥚 🐔 
https://dsdmoj.atlassian.net/browse/P4-1560
### What?

I have added/removed/altered:

- [x] Added a new filter to the moves endpoint

### Why?

To filter on moves with an allocation association, add a new filter
to either return all moves with an allocation if set to true, or to exclude
all moves associated to an allocation if set to false

### Have you? (optional)

- [x] Updated API docs if necessary	

### Deployment risks (optional)

- Changes an api that is used in production

### CODE REVIEW

File: app/services/moves/finder.rb [hunk 1]
```diff
@@ -84,5 +85,13 @@ def apply_supplier_filters(scope)
 
       scope.served_by(filter_params[:supplier_id])
     end
+
+    def apply_allocation_relationship_filters(scope)
+      return scope unless filter_params.key?(:has_relationship_to_allocation)
+
+      scope = scope.where.not(allocation: nil) if filter_params[:has_relationship_to_allocation] == 'true'
```
Comment (smoothcontract) [inline app/services/moves/finder.rb:92]: Could this be changed to use `allocation_id` rather than `allocation`? I'm thinking it might be a tad faster...

File: spec/requests/api/v1/moves_controller_filter_spec.rb [hunk 2]
```diff
@@ -142,6 +142,36 @@
...
+
+        it_behaves_like 'it returns the correct moves'
+      end
+
+      context 'when set to incorrect type' do
+        let(:expected_moves) { Move.where(id: [move_without_allocation.id, move_with_allocation.id]) }
+        let(:unexpected_moves) { Move.none }
+        let(:filter_value) { 'some-value' }
+
+        it_behaves_like 'it returns the correct moves'
```
Comment (smoothcontract) [inline spec/requests/api/v1/moves_controller_filter_spec.rb:171]: Does this need a test case for the key being excluded completely? Or set to `null` maybe?

File: spec/swagger/definitions/hand_coded_paths.yaml [hunk 3]
```diff
@@ -320,6 +320,13 @@
         items:
           type: string
           format: uuid
+    - name: filter[has_relationship_to_allocation]
+      in: query
+      explode: false
+      description: Filters results to either exclude or include moves associated to an allocation
```
Comment (smoothcontract) [inline spec/swagger/definitions/hand_coded_paths.yaml:326]: Maybe add the word `optional` in here just to clarify it can be omitted.
